app/context/authContext.tsx

```diff
@@ -0,0 +1,204 @@
+'use client';
+
+import React, { createContext, useContext, ReactNode, useMemo, useEffect } from 'react';
+import { useRouter } from 'next/navigation';
+import { MutationStatus, QueryStatus, useMutation, useQuery } from '@tanstack/react-query';
+import { clearSession, getSession, setSession } from '../../utils/actions/session';
+import getAuthUrl from '../../utils/getAuthUrl';
+
+const baseUrl = process.env.NEXT_PUBLIC_API_BASE_URL || '';
+
+export type Role = 'admin' | 'content_manager' | 'member';
+
+export type MiddlewareType = 'auth' | 'guest';
```

Comment:
I am not sure why we have the distinction here for auth and guest. Where would guest be applicable?

Reply:
Example: on the home page we want to show "Welcome {username}" , for that we would need to get the user object from the `useAuth` hook, but the home page can be viewed as guest as well, so in that case as a `middleware` value in the hook, we would set `guest` so the `useEffect` in the `useAuth` hook don't redirect the guest users to the `/login` route.